### PR TITLE
[activation checkpointing] Tagging based min cut partitioner

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #103601
* #102935
* __->__ #103357
* #102934

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @msaroufim @aakhundov

### CODE REVIEW

File: torch/_functorch/partitioners.py [hunk 1]
```diff
@@ -398,11 +587,21 @@ def min_cut_rematerialization_partition(
     fx_g = joint_module.graph
 
     #  add the CSE pass
-    if config.cse:
+    if config.cse and not config.functionalize_rng_ops:
```
Comment (anijain2305) [inline torch/_functorch/partitioners.py:590]: Clean this up

**Comment by anijain2305** [comment]
@jansel This is ready for review. @Chillee reviews this PR w/ me in my 1:1 and I have incorporated all his comments. His comments were - Avoid changing the existing code as much as possible, so that removing it later, if we need to, is easy.

File: torch/_functorch/partitioners.py [hunk 2]
```diff
@@ -26,6 +26,21 @@ def is_symint_node(node):
     assert hasattr(node, 'meta'), "All nodes traced with proxy_tensor should have meta"
     return "val" in node.meta and isinstance(node.meta['val'], torch.SymInt)
 
+def must_recompute(node):
+    return "recompute" in node.meta and node.meta["recompute"]
```
Comment (jansel) [inline torch/_functorch/partitioners.py:30]:
```suggestion
    return node.meta.get("recompute", False)
```

File: torch/_functorch/partitioners.py [hunk 3]
```diff
@@ -356,6 +372,241 @@ def pointwise_ops():
...
+    for node in list(gm.graph.nodes)[order[first_node_in_bwd]:]:
+        insert_node_in_graph(node)
+
+    # Build the output node
+    new_outputs = []
+    output_node = [node for node in gm.graph.nodes if node.op == "output"][0]
+    for output in output_node.args[0]:
+        if isinstance(output, torch.fx.node.Node):
+            if output not in env:
+                print("output gradient dependent only on fwd", output)
```
Comment (jansel) [inline torch/_functorch/partitioners.py:482]: print

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**